pf/internal/schemashim/attr_schema.go

```diff
@@ -86,39 +84,25 @@ func (*attrSchema) StateFunc() shim.SchemaStateFunc {
 
 // Needs to return a shim.Schema, a shim.Resource, or nil.
 func (s *attrSchema) Elem() interface{} {
-	t := s.attr.GetType()
-
+	switch t := s.attr.GetType().(type) {
 	// The ObjectType can be triggered through tfsdk.SingleNestedAttributes. Logically it defines an attribute with
 	// a type that is an Object type. To encode the schema of the Object type in a way the shim layer understands,
 	// Elem() needes to return a Resource value.
 	//
 	// See also: documentation on shim.Schema.Elem().
-	if tt, ok := t.(basetypes.ObjectTypable); ok {
-		var res shim.Resource = newObjectPseudoResource(tt, s.attr.Nested(), nil)
+	case basetypes.ObjectTypable:
+		var res shim.Resource = newObjectPseudoResource(t, s.attr.Nested(), nil)
 		return res
-	}
-	if tt, ok := t.(pfattr.TypeWithElementTypes); ok {
-		var res shim.Resource = newTuplePseudoResource(tt)
+	case pfattr.TypeWithElementTypes:
+		var res shim.Resource = newTuplePseudoResource(t)
 		return res
-	}
+	case pfattr.TypeWithElementType:
+		return shim.Schema(newTypeSchema(t.ElementType(), s.attr.Nested()))
```

Comment:
This looks reasonable, sounds like it subsumes all the Map/List/Set cases as well as might handle custom types that implement pfattr.TypeWithElementType interface without necessarily being one of the above concrete types. Custom collections?

Reply:
The change is necessitated to handle custom collections. This is the type that AWS introduced:

https://github.com/hashicorp/terraform-provider-aws/blob/3de2fd3ac7f5c9ddb7f64134fdc0dc5b7110fb79/internal/framework/types/list_nested_objectof.go#L17C1-L20

Reply:
Great fodder for a test case.